### PR TITLE
feat: add pre-quantized FP8 weight loading support

### LINKED ISSUE #3506: Support for loading pre-quantized FP8 models
### What feature would you like to be added?

Add support for loading pre-quantized FP8 models. For FP8 scaled_mm, the backend expects the weight tensor to either:

- originate from qfp8wt when it is produced on the device, or
- be transferred from the host to the device using the special DCI so that it has the expected layout/stickification.

### Why is this feature needed?

This feature is needed to support pre-quantized FP8 models, such as `granite-3.3-8b-instruct-fp8`. For these models, the pre-quantized weights should be transferred to the device using the expected stickification. Otherwise, the weights must first be quantized on the device using quantize_weight_fp8_with_scale. Without this support, pre-quantized FP8 models cannot be used because the backend expects the weight tensor to have the appropriate layout/stickification for FP8 scaled_mm.

### Any alternative solution that you may want to share?

No

### Anything else we need to know?

No

### PR DESCRIPTION
**Pull request by Akash-Nayak**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [  ] bug
- [x] feature
- [  ] documentation
- [  ] cleanup

#### What this PR does:

Enables end-to-end inference on pre-quantized FP8 model checkpoints (e.g.
`granite-3.3-8b-instruct-FP8`) on Spyre without any runtime dequantization and requantization.

Pre-quantized FP8 checkpoints store weights as `torch.float8_e4m3fn`. Without this PR, the only path onto Spyre was to cast those weights back to FP16 and re-quantize at runtime via the `qfp8wt` op — wasting memory bandwidth and defeating the purpose of shipping a pre-quantized checkpoint. This PR adds a direct KERNEL-layout DMA path that loads the FP8 bytes straight into the Spyre QFP8WT stick layout, and fixes the two bugs that blocked compilation for
real-world Granite shapes.

---

**1. Pre-quantized FP8 weight loading (`model_utils.py`)**

- `_dma_to_spyre_fp8_kernel()`: DMA helper that transfers a `float8_e4m3fn` weight tensor directly to Spyre as a KERNEL tensor with 2D stick `[si=2, so=64]` and `ElementArrangement.QFP8WT`, bypassing the runtime `qfp8wt` quantization op entirely.
- `use_fp8_weights` parameter on `load_model_to_spyre()`: when `True`, routes `float8_e4m3fn` `nn.Linear` weights through the FP8 KERNEL path (priority 1) before the normal FP16 path (priority 2).
- `load_fp8_model_to_spyre()`: convenience wrapper for pre-quantized FP8 model checkpoints.

**2. QFP8WT layout fixes (`propagate_layouts.py`, `work_division.py`)**

- Fix `_matmul_layouts` to skip `find_stick_compatible_input_layout` for the QFP8WT weight (it is a per-tile-fixed KERNEL tensor, not a normal stick input).
- Pin `_FP8_BMM_N_SPLIT = 4` for `batchmatmulfp8` N-dim work division.

**3. QFP8WT DCI stride interleave (`csrc/spyre_mem.cpp`)** — Bug fix

The QFP8WT physical layout `[si=2, so=64, K/si, N/so]` is a 4D stride-interleave that the generic `get_device_stride_infos()` path cannot express. Replaced with an explicit `DataConversionStrideInfo`:

```
stride_src = [N, 1, si*N, so]   # host row-major over [si, so, K/si, N/so]
stride_dst = [1, si, si*so, K*so]  # QFP8WT stick-packed device order
```

Without this fix, pre-quantized FP8 weights were silently
mis-transferred to device, producing wrong matmul results.

**4. `batchmatmulfp8` work division (`_inductor/work_division.py`)** — Bug fix

`batchmatmulfp8` was skipping the analytic cost model and landing on a fallback that did not enforce `n_dim % 4 == 0`. Fixed by routing through `_cost_model_divide_op` with `n_dim = _FP8_BMM_N_SPLIT = 4`, which resolves compilation aborts on non-4-divisible N shapes.

---

**Tests**

- `test_fp8_prequantized_load.py` (new, 20 tests): QFP8WT DMA transfer and layout validation; pre-quantized weight `_scaled_mm` correctness; `load_fp8_model_to_spyre()` with `nn.Linear` FP8 weights.
- `test_copy_from_d2d_offsets.py` (new, 14 tests): QFP8WT KERNEL tensor clone regressions; D2D offset correctness; strided view coverage.
- `test_inductor_ops.py -k fp8` (58 tests): all Granite 8B shapes pass (`fused_qkv N=6144`, `fused_gate_up N=25600`, wide-M prefill/decode).

All 164 tests pass, 7 intentional `xfail`s.

#### Which issue(s) this PR is related to:

Fixes #3506 (pre-quantized FP8 weight support)

#### Special notes for your reviewer:

- The DCI stride fix (item 3) and the work-division fix (item 4) are independent of each other but both required for correct end-to-end execution at Granite 8B shapes.
- `TestCopyFromD2DFP8QFP8WT::test_column_tile_clone` requires a deeptools compiler fix ([deeptools#4687](https://github.ibm.com/ai-chip-toolchain/deeptools/issues/4687), merged via [deeptools#4689](https://github.ibm.com/ai-chip-toolchain/deeptools/pull/4689)) — not part of this PR. The test were run locally with `DEEPTOOLS_PATH=/path/to/patched/deeptools` with the patch specified in the deeptools issue.
- `TestGraniteF8IntegrationLoad` (full model load from NFS) has been excluded from this PR — it requires `transformers`, `safetensors`, and the NFS model mount which are not available in standard CI. The remaining tests are fully self-contained.
- `test_scaled_mm_prequantized_weight_xfail` cases document the intentionally unsupported "KERNEL weight as graph input" pattern and are marked `xfail(strict=True)`.

#### Does this PR introduce a user-facing change?

Yes. Adds `load_fp8_model_to_spyre()` public API and `use_fp8_weights` parameter to `load_model_to_spyre()` for loading pre-quantized FP8 model checkpoints directly onto Spyre without runtime requantization.

Co-authored-by: Avery Blanchard <avery.blanchard@ibm.com>
Co-authored-by: AskBob <askbob@ibm.com>

### CODE REVIEW

File: tests/inductor/test_fp8_prequantized_load.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+        assert q_weight.shape == torch.Size([n, k]), (
+            f"Expected shape [{n}, {k}], got {q_weight.shape}"
+        )
+
+    @pytest.mark.xfail(
+        reason=(
+            "Weight as compiled-graph input causes device mismatch (cpu vs spyre:0) "
+            "during tracing. Pre-loaded KERNEL weights must be closed over as frozen "
+            "constants, not passed as graph inputs. See class docstring."
+        ),
```
Comment (avery-blanchard) [inline tests/inductor/test_fp8_prequantized_load.py:245]: Please correct me if I am missing something here. Is this not an issue with how the test is written? After `_dma_to_spyre_fp8_kernel`, `q_weight` is a spyre/device tensor and it is being passed as a argument to the CPU function?

Comment (Akash-Nayak) [reply tests/inductor/test_fp8_prequantized_load.py:245]: You're right, good catch. `compare_with_pytorch` passes the same arguments to both the compiled Spyre path and the CPU reference. So, `q_weight` that is already on `spyre:0` after `_dma_to_spyre_fp8_kernel` was also being passed to `fn_pytorch`, where `q_weight.to(torch.float32)` expects a CPU tensor. This was an incidental CPU-reference failure on top of the actual Spyre tracing issue being documented.

Fixed in `ec51a0ed`: `pytorch_fn` now closes over the original CPU `weight_fp8` tensor instead of taking `q_weight` as an argument. This isolates the `xfail` to the actual Spyre-side failure: Dynamo treats the pre-loaded `spyre:0` tensor as a graph input rather than a frozen constant, which leads to incorrect results.

Comment (avery-blanchard) [reply tests/inductor/test_fp8_prequantized_load.py]: Could you clarify the issue with this being a graph input?

Comment (Akash-Nayak) [reply tests/inductor/test_fp8_prequantized_load.py]: You were right to question this. After investigating further, the original xfail was incorrect.

I tested the graph-input and closed-over cases with the correctly prepared KERNEL weight, and they produce bit-identical hardware output (0.0 max difference across the tested shapes). So there is no issue with passing a pre-loaded QFP8WT KERNEL tensor as a graph input.

The mismatch I initially observed was caused by the test setup: `_dma_to_spyre_fp8_kernel` expects the weight in `[k, n]` orientation, whereas the test was passing the `[n, k]` weight directly. The other prequantized tests use `.T.contiguous()` before the DMA call, which this test was missing.

I have fixed that in `170df30a`, removed the xfail, and renamed the test to `test_scaled_mm_prequantized_weight_as_graph_input`. I also corrected the CPU reference and aligned the tolerance with the existing closed-over prequantized test.

With those changes, all 22 tests in `test_fp8_prequantized_load.py` and the relevant 130 FP8/scaled_mm tests pass in test_inductor_ops.py and test_fp8_operations.py.
